File: Appointment/AppointmentStore.m
#import "AppointmentStore.h"

@implementation AppointmentStore

+ (AppointmentStore *)shared {
	static AppointmentStore *shared = nil;
	
	@synchronized(self) {
		if (shared == nil) {
			shared = [[self alloc] init];
		}
	}
	
	return shared;
}

- (instancetype)init {
	self = [super init];
	
	if (self) {
		[self setUpContext];
	}
	
	return self;
}

- (void)setUpContext {
	model = [NSManagedObjectModel mergedModelFromBundles:nil];
	
	NSPersistentStoreCoordinator *psc = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:model];
	
	NSString *documentDirectory = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
	
	NSURL *storeURL = [NSURL fileURLWithPath:[documentDirectory stringByAppendingPathComponent:@"store.data"]];
	
	NSError *error = nil;
	
	if (![psc addPersistentStoreWithType:NSSQLiteStoreType
						   configuration:nil
									 URL:storeURL
								 options:@{NSMigratePersistentStoresAutomaticallyOption:@ YES, NSInferMappingModelAutomaticallyOption:@ YES}
								   error: &error]) {
		
		[NSException raise: @"Open failed" format:@"Reason: %@", error.localizedDescription];
	}
	
	objectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
	objectContext.persistentStoreCoordinator = psc;
}

- (void)saveChanges {
	NSError *error = nil;
	
	if (objectContext.hasChanges) {
		[objectContext save:&error];
		[objectContext performBlockAndWait:^{
			if (error) {
				NSLog(@"Error saving: %@", error.localizedDescription);
			}
		}];
	}
}

- (void)removeAppointment:(Appointment *)appointment {
	[objectContext deleteObject:appointment];
}

- (void)removeAllAppointments {
	NSArray *stores = objectContext.persistentStoreCoordinator.persistentStores;
	
	for (NSPersistentStore *store in stores) {
		[objectContext.persistentStoreCoordinator removePersistentStore:store error:nil];
		[[NSFileManager defaultManager] removeItemAtPath:store.URL.path error:nil];
	}
	
	[self setUpContext];
}

- (NSArray *)allAppointments {
	NSFetchRequest *request = [[NSFetchRequest alloc] init];
	
	NSEntityDescription *entity = model.entitiesByName[@"Appointment"];
	request.entity = entity;
	
	NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"appointmentTime" ascending:YES];
	request.sortDescriptors = @[sortDescriptor];
	
	NSError *error;
	NSArray *result = [objectContext executeFetchRequest:request error:&error];
	
	if (!result) {
		[NSException raise:@"Fetch failed" format:@"Reason: %@", error.localizedDescription];
	}
	
	return [[NSMutableArray alloc] initWithArray:result];
}

- (Appointment *)createAppointment {
	Appointment *newAppointment = [NSEntityDescription insertNewObjectForEntityForName:@"Appointment" inManagedObjectContext:objectContext];
	
	return newAppointment;
}

@end

File: Appointment/AppointmentStore.swift
//
//  AppointmentStore.swift
//  Appointment
//
//  Created by Jeffrey Moran on 10/26/20.
//  Copyright © 2020 Jeff Moran. All rights reserved.
//

import CoreData
import Foundation

class AppointmentStore: NSObject {

    // MARK: - Static Properties

    @objc static var shared = AppointmentStore()

    // MARK: - Private Properties

    private lazy var storeCoordinator: NSPersistentStoreCoordinator = {
        guard let model = NSManagedObjectModel.mergedModel(from: nil) else {
            fatalError("Model not created?")
        }

        let storeCoordinator = NSPersistentStoreCoordinator(managedObjectModel: model)

        let path = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first
        let url = URL(fileURLWithPath: path ?? "").appendingPathComponent("store.data")

        do {
            let options = [
                NSMigratePersistentStoresAutomaticallyOption: true,
                NSInferMappingModelAutomaticallyOption: true
            ]

            try storeCoordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: url, options: options)
        } catch {
            fatalError(error.localizedDescription)
        }

        return storeCoordinator
    }()

    private lazy var objectContext: NSManagedObjectContext = {
        let context = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
        context.persistentStoreCoordinator = storeCoordinator

        return context
    }()

    // MARK: - Internal Properties

    var count: Int {
        do {
            return try objectContext.count(for: Appointment.fetchRequest() as NSFetchRequest<NSFetchRequestResult>)
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    var allAppointments: [Appointment] {
        #warning("Add ability to fetch with sort")

        do {
            // swiftlint:disable force_cast
            return try objectContext.fetch(Appointment.fetchRequest() as NSFetchRequest<NSFetchRequestResult>) as! [Appointment]
            // swiftlint:enable force_cast
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    @objc var emptyAppointment: Appointment? {
        let appointment = NSEntityDescription.insertNewObject(forEntityName: "Appointment", into: objectContext)

        return appointment as? Appointment
    }

    // MARK: - Internal Methods

    @objc func save() {
        do {
            if objectContext.hasChanges {
                try objectContext.save()
            }
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    func deleteAll() {
        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = Appointment.fetchRequest()
        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)

        do {
            try objectContext.execute(deleteRequest)
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    func delete(_ appointment: Appointment) {
        objectContext.delete(appointment)
    }
}
